### DISCUSSION TITLE
Switch logging to "debug" lib?

### DISCUSSION THREAD

**Discussion by Apollon77** [mfucci/node-matter#65] [General]
Hi,

should we move from the console.log logs right now to "debug" lib?

Would have the benefit that loggign has time and timings and that we can structure logs by different sections of the flow. And it is not there by default but easy to enable.

**Comment by mfucci** [comment]
I do agree for the need of cleaner logging framework, I was about to add this.

However I was currently trying to stick to a "zero-dependency" philosophy.
Reasons:

- keep the codebase minimal
- simplify short / long term maintenance: asking authors of external libraries to fix bugs in their code is time-consumption and sometime end up in forking the library
- simplify the support of more environment: right now it is node-only, but I will add support for browser and mobile webviews, someone is working on supporting Espriuno.

Obviously, the main tradeoff is more code to write and maintain...

I am still planning on spending some time to get rid of the two current dependencies. if possible (but they are kind of hard to rewrite...).

A logging system doesn't seem to take much more than 100 lines of code, so not sure that it is worth linking to an external library for this.

Basic features of a logging system should be:

- debug / info / warn / error / fatal level
- debug / info level can be pruned by the minification system so they don't have any impact on production version performance
- scoped logs, level can be change per module in an easy configuration
- timestamp on logs

Anything else?

Maybe worth creating a specific workstream to implement the logging system and discuss features / libraries.

**Reply by Apollon77** [reply]
Ok, if this is your goal then you you most likely need to write something yourself.

To the features:
* I think error is enough and fatal not needed, but thats "flavour"
* Removing (debug/info) logging by "minification system " should be optional, because (at least in the very beginning of the whole matter topic - and from my experience also later - we will debug a lot in the productuon environments where the devices are ;-) Yes that would be a "nice to have for later when Matter as tech has stabilized" in ym eyes

Rest of your list should it already be, you're right, maybe beside:
* allow to hand in own logger methods and not pin to "console.log" to allow to provide own logger implementations

**Comment by mfucci** [comment]
Here is a PR adding a logger: https://github.com/mfucci/node-matter/pull/73
Let me know if this seems good in term of features.

It is only 65 lines of code :-)
